// grammarkdown/dist/emitter/ecmarkup.js
"use strict";
/*!
 * Copyright (c) 2020 Ron Buckton (rbuckton@chronicles.org)
 *
 * This file is licensed to you under the terms of the MIT License, found in the LICENSE file
 * in the root of this repository or package.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.EcmarkupEmitter = void 0;
const emitter_1 = require("./emitter");
const tokens_1 = require("../tokens");
/** {@docCategory Emit} */
class EcmarkupEmitter extends emitter_1.Emitter {
    constructor() {
        super(...arguments);
        this.extension = ".emu.html";
    }
    emitProduction(node) {
        const linkId = this.resolver.getProductionLinkId(node.name);
        this.emitLinkAnchor(linkId);
        this.writer.write(`<emu-production name="`);
        this.emitIdentifier(node.name);
        this.writer.write(`"`);
        this.emitNode(node.parameterList);
        if (node.colonToken) {
            switch (node.colonToken.kind) {
                case tokens_1.SyntaxKind.ColonColonToken:
                    this.writer.write(` type="lexical"`);
                    break;
                case tokens_1.SyntaxKind.ColonColonColonToken:
                    this.writer.write(` type="regexp"`);
                    break;
            }
        }
        if (node.body) {
            if (node.body.kind === tokens_1.SyntaxKind.OneOfList) {
                this.writer.write(` oneof`);
            }
            else if (node.body.kind === tokens_1.SyntaxKind.RightHandSide) {
                this.writer.write(` collapsed`);
            }
        }
        this.writer.write(`>`);
        this.writer.indent();
        this.writer.writeln();
        this.emitNode(node.body);
        this.writer.dedent();
        this.writer.writeln();
        this.writer.write(`</emu-production>`);
    }
    emitParameterList(node) {
        this.writer.write(` params="`);
        if (node.elements) {
            for (let i = 0; i < node.elements.length; ++i) {
                if (i > 0) {
                    this.writer.write(`, `);
                }
                this.emitNode(node.elements[i]);
            }
        }
        this.writer.write(`"`);
    }
    emitParameter(node) {
        this.emitIdentifier(node.name);
    }
    emitOneOfList(node) {
        this.writer.write(`<emu-rhs>`);
        if (node.terminals) {
            for (let i = 0; i < node.terminals.length; ++i) {
                const terminal = node.terminals[i];
                this.emitDetachedTriviaOfNode(terminal);
                if (i > 0) {
                    this.writer.write(` `);
                }
                this.emitLeadingTriviaOfNode(terminal);
                this.emitTextContent(terminal);
                this.emitTrailingTriviaOfNode(terminal);
            }
        }
        this.writer.write(`</emu-rhs>`);
    }
    emitRightHandSideList(node) {
        if (node.elements) {
            for (const rhs of node.elements) {
                this.beforeEmitNode(rhs);
                this.emitRightHandSideWorker(rhs, /*inList*/ true);
                this.afterEmitNode(rhs);
            }
        }
    }
    emitRightHandSide(node) {
        this.emitRightHandSideWorker(node, /*inList*/ false);
    }
    emitRightHandSideWorker(node, inList) {
        const linkId = this.resolver.getRightHandSideLinkId(node, /*includePrefix*/ false);
        this.emitLinkAnchor(linkId);
        this.writer.write(`<emu-rhs`);
        if (linkId) {
            this.writer.write(` a="${linkId}"`);
        }
        if (node.constraints) {
            this.writer.write(` constraints="`);
            this.emitNode(node.constraints);
            this.writer.write(`"`);
        }
        this.writer.write(`>`);
        if (inList || node.head && node.head.next) {
            this.writer.indent();
            this.writer.writeln();
            this.emitNode(node.head);
            this.writer.dedent();
            this.writer.writeln();
        }
        else {
            this.emitNode(node.head);
        }
        this.writer.write(`</emu-rhs>`);
    }
    afterEmitNode(node) {
        super.afterEmitNode(node);
        switch (node.kind) {
            case tokens_1.SyntaxKind.RightHandSideList:
            case tokens_1.SyntaxKind.RightHandSide:
            case tokens_1.SyntaxKind.OneOfList:
            case tokens_1.SyntaxKind.Production:
                this.writer.writeln();
                break;
        }
    }
    emitSymbolSpan(node) {
        this.emitNode(node.symbol);
        if (node.next) {
            this.writer.writeln();
            this.emitNode(node.next);
        }
    }
    emitPlaceholder(node) {
        this.writer.write(`<emu-t placeholder>@</emu-t>`);
    }
    emitTerminal(node) {
        const isUnicodeCharacter = node.literal.kind === tokens_1.SyntaxKind.UnicodeCharacterLiteral;
        this.writer.write(isUnicodeCharacter ? `<emu-gprose` : `<emu-t`);
        if (node.questionToken) {
            this.writer.write(` optional`);
        }
        this.writer.write(`>`);
        this.emitTextContent(node.literal);
        this.writer.write(isUnicodeCharacter ? `</emu-gprose>` : `</emu-t>`);
    }
    emitNonterminal(node) {
        const linkId = this.resolver.getProductionLinkId(node.name);
        this.writer.write(`<emu-nt`);
        this.emitNode(node.argumentList);
        if (node.questionToken) {
            this.writer.write(` optional`);
        }
        this.writer.write(`>`);
        this.emitNodeWithLink(node.name, linkId);
        this.writer.write(`</emu-nt>`);
    }
    emitArgumentList(node) {
        this.writer.write(` params="`);
        if (node.elements) {
            for (let i = 0; i < node.elements.length; ++i) {
                if (i > 0) {
                    this.writer.write(`, `);
                }
                this.emitNode(node.elements[i]);
            }
        }
        this.writer.write(`"`);
    }
    emitArgument(node) {
        this.emitToken(node.operatorToken);
        this.emitNode(node.name);
    }
    emitTerminalLiteral(node) {
        this.writer.write(`<emu-t>`);
        this.emitTextContent(node);
        this.writer.write(`</emu-t>`);
    }
    emitUnicodeCharacterRange(node) {
        this.writer.write(`<emu-gprose>`);
        this.emitTextContent(node.left);
        this.writer.write(` through `);
        this.emitTextContent(node.right);
        this.writer.write(`</emu-gprose>`);
    }
    emitUnicodeCharacterLiteral(node) {
        this.writer.write(`<emu-gprose>`);
        this.emitTextContent(node);
        this.writer.write(`</emu-gprose>`);
    }
    emitProse(node) {
        this.writer.write(`<emu-gprose>`);
        if (node.fragments) {
            for (const fragment of node.fragments) {
                this.emitNode(fragment);
            }
        }
        this.writer.write(`</emu-gprose>`);
    }
    emitEmptyAssertion(node) {
        this.writer.write(`<emu-gann>empty</emu-gann>`);
    }
    emitSymbolSet(node) {
        this.writer.write(`{`);
        if (node.elements) {
            for (let i = 0; i < node.elements.length; ++i) {
                if (i > 0) {
                    this.writer.write(`,`);
                }
                this.writer.write(` `);
                this.emitNode(node.elements[i]);
            }
        }
        this.writer.write(` }`);
    }
    emitLookaheadAssertion(node) {
        this.writer.write(`<emu-gann>`);
        if (node.operatorToken) {
            switch (node.operatorToken.kind) {
                case tokens_1.SyntaxKind.ExclamationEqualsToken:
                case tokens_1.SyntaxKind.NotEqualToToken:
                    this.writer.write(`lookahead ≠ `);
                    break;
                case tokens_1.SyntaxKind.EqualsToken:
                case tokens_1.SyntaxKind.EqualsEqualsToken:
                    this.writer.write(`lookahead = `);
                    break;
                case tokens_1.SyntaxKind.ElementOfToken:
                case tokens_1.SyntaxKind.LessThanMinusToken:
                    this.writer.write(`lookahead ∈ `);
                    break;
                case tokens_1.SyntaxKind.NotAnElementOfToken:
                case tokens_1.SyntaxKind.LessThanExclamationToken:
                    this.writer.write(`lookahead ∉ `);
                    break;
            }
        }
        this.emitNode(node.lookahead);
        this.writer.write(`</emu-gann>`);
    }
    emitLexicalGoalAssertion(node) {
        this.writer.write(`<emu-gann>lexical goal `);
        const linkId = node.symbol && this.resolver.getProductionLinkId(node.symbol);
        this.emitNodeWithLink(node.symbol, linkId);
        this.writer.write(`</emu-gann>`);
    }
    emitNoSymbolHereAssertion(node) {
        this.writer.write(`<emu-gann>no `);
        if (node.symbols) {
            for (let i = 0; i < node.symbols.length; ++i) {
                if (i > 0) {
                    this.writer.write(` or `);
                }
                this.emitNode(node.symbols[i]);
            }
        }
        this.writer.write(` here</emu-gann>`);
    }
    emitConstraints(node) {
        if (node.elements) {
            for (let i = 0; i < node.elements.length; ++i) {
                if (i > 0) {
                    this.writer.write(`, `);
                }
                this.emitNode(node.elements[i]);
            }
        }
    }
    emitProseAssertion(node) {
        this.writer.write(`<emu-gmod>`);
        if (node.fragments) {
            for (const fragment of node.fragments) {
                this.emitNode(fragment);
            }
        }
        this.writer.write(`</emu-gmod>`);
    }
    emitButNotSymbol(node) {
        this.emitNode(node.left);
        this.writer.write(` <emu-gmod>but not `);
        this.emitNode(node.right);
        this.writer.write(`</emu-gmod>`);
    }
    emitOneOfSymbol(node) {
        this.writer.write(`one of `);
        if (node.symbols) {
            for (let i = 0; i < node.symbols.length; ++i) {
                if (i > 0) {
                    this.writer.write(` or `);
                }
                this.emitNode(node.symbols[i]);
            }
        }
    }
    emitTextContent(node) {
        if (node && node.text) {
            this.writer.write(this.encode(node.text));
        }
    }
    emitLinkAnchor(linkId) {
        if (linkId && this.options.emitLinks) {
            this.writer.write(`<a name="${linkId}"></a>`);
        }
    }
    emitNodeWithLink(node, linkId) {
        if (linkId && this.options.emitLinks) {
            this.writer.write(`<a href="#${linkId}">`);
            this.emitNode(node);
            this.writer.write(`</a>`);
        }
        else {
            this.emitNode(node);
        }
    }
}
exports.EcmarkupEmitter = EcmarkupEmitter;
