// grammarkdown/dist/index.js
"use strict";
/*!
 * Copyright (c) 2020 Ron Buckton (rbuckton@chronicles.org)
 *
 * This file is licensed to you under the terms of the MIT License, found in the LICENSE file
 * in the root of this repository or package.
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.getDefaultOptions = exports.NewLineKind = exports.EmitFormat = exports.NodeAsyncHost = exports.StringAsyncHost = exports.CoreAsyncHost = exports.Range = exports.Position = void 0;
var types_1 = require("./types");
Object.defineProperty(exports, "Position", { enumerable: true, get: function () { return types_1.Position; } });
Object.defineProperty(exports, "Range", { enumerable: true, get: function () { return types_1.Range; } });
var host_1 = require("./host");
Object.defineProperty(exports, "CoreAsyncHost", { enumerable: true, get: function () { return host_1.CoreAsyncHost; } });
Object.defineProperty(exports, "StringAsyncHost", { enumerable: true, get: function () { return host_1.StringAsyncHost; } });
var node_1 = require("./hosts/node");
Object.defineProperty(exports, "NodeAsyncHost", { enumerable: true, get: function () { return node_1.NodeAsyncHost; } });
__exportStar(require("./diagnostics"), exports);
var options_1 = require("./options");
Object.defineProperty(exports, "EmitFormat", { enumerable: true, get: function () { return options_1.EmitFormat; } });
Object.defineProperty(exports, "NewLineKind", { enumerable: true, get: function () { return options_1.NewLineKind; } });
Object.defineProperty(exports, "getDefaultOptions", { enumerable: true, get: function () { return options_1.getDefaultOptions; } });
__exportStar(require("./tokens"), exports);
__exportStar(require("./nodes"), exports);
__exportStar(require("./symbols"), exports);
__exportStar(require("./scanner"), exports);
__exportStar(require("./parser"), exports);
__exportStar(require("./binder"), exports);
__exportStar(require("./checker"), exports);
__exportStar(require("./emitter"), exports);
__exportStar(require("./stringwriter"), exports);
__exportStar(require("./grammar"), exports);
__exportStar(require("./navigator"), exports);
__exportStar(require("./visitor"), exports);
