// grammarkdown/dist/nodeInternal.js
"use strict";
/*!
 * Copyright (c) 2021 Ron Buckton (rbuckton@chronicles.org)
 *
 * This file is licensed to you under the terms of the MIT License, found in the LICENSE file
 * in the root of this repository or package.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.getSourceFileAccessor = exports.setSourceFileAccessor = exports.getTriviaAccessor = exports.setTriviaAccessor = exports.getNodeAccessor = exports.setNodeAccessor = void 0;
let nodeAccessor;
let triviaAccessor;
let sourceFileAccessor;
/* @internal */
function setNodeAccessor(accessor) { nodeAccessor = accessor; }
exports.setNodeAccessor = setNodeAccessor;
/* @internal */
function getNodeAccessor() {
    if (!nodeAccessor)
        throw new TypeError("Accessor not set");
    return nodeAccessor;
}
exports.getNodeAccessor = getNodeAccessor;
/* @internal */
function setTriviaAccessor(accessor) { triviaAccessor = accessor; }
exports.setTriviaAccessor = setTriviaAccessor;
/* @internal */
function getTriviaAccessor() {
    if (!triviaAccessor)
        throw new TypeError("Accessor not set");
    return triviaAccessor;
}
exports.getTriviaAccessor = getTriviaAccessor;
/* @internal */
function setSourceFileAccessor(accessor) { sourceFileAccessor = accessor; }
exports.setSourceFileAccessor = setSourceFileAccessor;
/* @internal */
function getSourceFileAccessor() {
    if (!sourceFileAccessor)
        throw new TypeError("Accessor not set");
    return sourceFileAccessor;
}
exports.getSourceFileAccessor = getSourceFileAccessor;
