// ecmarkdown/dist/visitor.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const childKeys = {
    opaqueTag: [],
    tag: [],
    comment: [],
    algorithm: ['contents'],
    text: [],
    'double-brackets': [],
    star: ['contents'],
    underscore: [],
    tick: ['contents'],
    tilde: ['contents'],
    pipe: [],
    ul: ['contents'],
    ol: ['contents'],
    'ordered-list-item': ['contents', 'sublist'],
    'unordered-list-item': ['contents', 'sublist'],
};
function visit(node, observer) {
    var _a, _b;
    if (Array.isArray(node)) {
        node.forEach(n => visit(n, observer));
        return;
    }
    (_a = observer.enter) === null || _a === void 0 ? void 0 : _a.call(observer, node);
    // @ts-ignore
    for (let childKey of childKeys[node.name]) {
        // @ts-ignore
        let child = node[childKey];
        if (child === null) {
            continue;
        }
        if (Array.isArray(child)) {
            for (let c of child) {
                visit(c, observer);
            }
        }
        else {
            visit(child, observer);
        }
    }
    (_b = observer.exit) === null || _b === void 0 ? void 0 : _b.call(observer, node);
}
exports.visit = visit;
